### PR TITLE
Sync v4 with latest from master and develop

### PR DESCRIPTION
**Pull request by svlandeg**
Don't squash merge.

## Description
Sync v4 branch

### Types of change
chore

## Checklist
- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

**Review by rmitsch — APPROVED** [review]
I didn't see any issues when going through the changes.

File: spacy/ml/tb_framework.pyx [hunk 1]
```diff
@@ -145,7 +145,7 @@ def init(
         current_nO = model.maybe_get_dim("nO")
         if current_nO is None or current_nO != inferred_nO:
             model.attrs["resize_output"](model, inferred_nO)
-    nO = model.get_dim("nO")
+    # nO = model.get_dim("nO")
```
Comment (adrianeboyd) [inline spacy/ml/tb_framework.pyx:148]:
```suggestion
```

File: spacy/ml/tb_framework.pyx [hunk 2]
```diff
@@ -302,8 +304,8 @@ def _forward_fallback(
     backprop_feats,
     seen_mask,
     is_train: bool,
-    actions: Optional[List[Ints1d]]=None,
-    max_moves: int=0):
+    actions: Optional[List[Ints1d]] = None,
+        max_moves: int = 0):
```
Comment (adrianeboyd) [inline spacy/ml/tb_framework.pyx:308]: This indent seems strange?

Comment (rmitsch) [inline spacy/ml/tb_framework.pyx:308]:
```suggestion
    actions: Optional[List[Ints1d]] = None,
    max_moves: int = 0
):
```

Comment (rmitsch) [reply spacy/ml/tb_framework.pyx:308]: Right, probably a mistake in my linting fixes. See https://github.com/explosion/spaCy/pull/12842#discussion_r1269043556.

**Pull request merged**